src/free_functions.rs

```diff
@@ -64,32 +65,105 @@ pub fn rcarr1<A: Clone>(xs: &[A]) -> ArcArray1<A> {
 }
 
 /// Create a zero-dimensional array view borrowing `x`.
-pub fn aview0<A>(x: &A) -> ArrayView0<'_, A> {
-    unsafe { ArrayView::from_shape_ptr(Ix0(), x) }
+pub const fn aview0<A>(x: &A) -> ArrayView0<'_, A> {
+    ArrayBase {
+        data: ViewRepr::new(),
+        // Safe because references are always non-null.
+        ptr: unsafe { NonNull::new_unchecked(x as *const A as *mut A) },
+        dim: Ix0(),
+        strides: Ix0(),
+    }
 }
 
 /// Create a one-dimensional array view with elements borrowing `xs`.
 ///
+/// **Panics** if the length of the slice overflows `isize`. (This can only
+/// occur if `A` is zero-sized, because slices cannot contain more than
+/// `isize::MAX` number of bytes.)
+///
 /// ```
-/// use ndarray::aview1;
+/// use ndarray::{aview1, ArrayView1};
 ///
 /// let data = [1.0; 1024];
 ///
 /// // Create a 2D array view from borrowed data
 /// let a2d = aview1(&data).into_shape((32, 32)).unwrap();
 ///
 /// assert_eq!(a2d.sum(), 1024.0);
+///
+/// // Create a const 1D array view
+/// const C: ArrayView1<'static, f64> = aview1(&[1., 2., 3.]);
+///
+/// assert_eq!(C.sum(), 6.);
 /// ```
-pub fn aview1<A>(xs: &[A]) -> ArrayView1<'_, A> {
-    ArrayView::from(xs)
+pub const fn aview1<A>(xs: &[A]) -> ArrayView1<'_, A> {
+    if size_of::<A>() == 0 {
+        assert!(
+            xs.len() <= isize::MAX as usize,
+            "Slice length must fit in `isize`.",
+        );
+    }
+    ArrayBase {
```

Comment:
is it possible to use the internal builder methods, with_data_ptr etc? We'd prefer those instead of spreading out `ArrayBase { .. }` constructors again.

Reply:
If I try to make `from_data_ptr` into a const fn, then I get the following compilation errors:

```
error[E0658]: trait bounds other than `Sized` on const fn parameters are unstable
  --> src/impl_internal_constructors.rs:14:9
   |
14 | impl<A, S> ArrayBase<S, Ix1>
   |         ^
...
26 |     pub(crate) const unsafe fn from_data_ptr(data: S, ptr: NonNull<A>) -> Self {
   |     -------------------------------------------------------------------------- function declared as const here
   |
   = note: see issue #57563 <https://github.com/rust-lang/rust/issues/57563> for more information

error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
  --> src/impl_internal_constructors.rs:33:23
   |
33 |         debug_assert!(array.pointer_is_inbounds());
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0493]: destructors cannot be evaluated at compile-time
  --> src/impl_internal_constructors.rs:27:13
   |
27 |         let array = ArrayBase {
   |             ^^^^^ constant functions cannot evaluate destructors
...
35 |     }
   |     - value is dropped here
```

The biggest issue is the generic `S` type. I don't see a way to fix the errors while keeping a generic `S`. We could add const constructors to `impl_internal_constructors.rs` specifically for `ArrayView`, though.